code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/vnext/common/AEPPresentable.kt

```diff
@@ -0,0 +1,328 @@
+    ) : this(
+        presentation,
+        presentationUtilityProvider,
+        presentationDelegate,
+        appLifecycleProvider,
+        PresentationStateManager(),
+        CoroutineScope(Dispatchers.Main)
+    )
+
+    /**
+     * @param presentation the [Presentation] to be used by this [Presentable]
+     * @param presentationUtilityProvider the [PresentationUtilityProvider] to be used to fetch components needed by this [Presentable]
+     * @param presentationDelegate the [PresentationDelegate] for notifying the application of [Presentation] lifecycle events
+     * @param appLifecycleProvider the [AppLifecycleProvider] to be used to register for app lifecycle events
+     * @param presentationStateManager the [PresentationStateManager] to be used to manage the state of this [Presentable]
+     * @param mainScope the [CoroutineScope] to be used to launch coroutines on the main thread
+     */
+    @VisibleForTesting
+    internal constructor(
+        presentation: Presentation<T>,
+        presentationUtilityProvider: PresentationUtilityProvider,
+        presentationDelegate: PresentationDelegate?,
+        appLifecycleProvider: AppLifecycleProvider,
+        presentationStateManager: PresentationStateManager,
+        mainScope: CoroutineScope
+    ) {
+        this.presentation = presentation
+        this.presentationUtilityProvider = presentationUtilityProvider
+        this.presentationDelegate = presentationDelegate
+        this.appLifecycleProvider = appLifecycleProvider
+        this.presentationStateManager = presentationStateManager
+        this.mainScope = mainScope
+    }
+
+    override fun show() {
+        mainScope.launch {
+            val currentState = getState()
+            if (currentState == Presentable.State.VISIBLE) {
+                Log.debug(
+                    ServiceConstants.LOG_TAG,
+                    LOG_SOURCE,
+                    "Presentable is already shown. Ignoring show request."
+                )
+                return@launch
+            }
+
+            val currentActivity: Activity? = presentationUtilityProvider.getCurrentActivity()
+            if (currentActivity == null) {
+                Log.debug(
+                    ServiceConstants.LOG_TAG,
+                    LOG_SOURCE,
+                    "Current activity is null. Cannot show presentable."
+                )
+                return@launch
+            }
+
+            // If all basic conditions are met, check with the delegate if the presentable can be shown
+            if (gateDisplay()) { // TODO :should also include PresentationObserver gate
+                val canShow = (presentationDelegate?.canShow(this@AEPPresentable) ?: true)
+                if (!canShow) return@launch
+            }
+
+            // Show the presentable on the current activity
+            show(currentActivity)
+            // Register for app lifecycle events
+            appLifecycleProvider.registerListener(this@AEPPresentable)
```

Comment:
Transition from hide() -> show() will register the listener multiple times.

Reply:
AppLifecycleProvider maintains a set of listeners so registering and unregistering multiple times will result in registration and de-registration just once.

Reply:
I did not notice that it maintains a Set of listeners. Can you add a comment explaining it?